v1/ocagent.go

```diff
@@ -0,0 +1,282 @@
+	e.nodeInfo = createNodeInfo(e.serviceName)
+	return e, nil
+}
+
+const (
+	maxInitialConfigRetries = 10
+	maxInitialTracesRetries = 10
+)
+
+// Start dials to the agent, establishing a connection to it. It also
+// initiates the Config and Trace services by sending over the initial
+// messages that consist of the node identifier. Start performs a best case
+// attempt to try to send the initial messages, by applying exponential
+// backoff at most 10 times.
+func (ae *Exporter) Start() error {
+	ae.mu.Lock()
+	defer ae.mu.Unlock()
+
+	err := ae.doStartLocked()
+	if err == nil {
+		ae.started = true
+		return nil
+	}
+
+	// Otherwise we have an error and should clean up to avoid leaking resources.
+	ae.started = false
+	if ae.grpcClientConn != nil {
+		ae.grpcClientConn.Close()
+	}
+
+	return err
+}
+
+func (ae *Exporter) prepareAgentAddress() string {
+	if ae.agentAddress != "" {
+		return ae.agentAddress
+	}
+	port := DefaultAgentPort
+	if ae.agentPort > 0 {
+		port = ae.agentPort
+	}
+	return fmt.Sprintf("%s:%d", DefaultAgentHost, port)
+}
+
+func (ae *Exporter) doStartLocked() error {
+	if ae.started {
+		return nil
+	}
+
+	// Now start it
+	addr := ae.prepareAgentAddress()
+	var dialOpts []grpc.DialOption
+	if ae.canDialInsecure {
+		dialOpts = append(dialOpts, grpc.WithInsecure())
+	}
+
+	cc, err := grpc.Dial(addr, dialOpts...)
+	if err != nil {
+		return err
+	}
+	ae.grpcClientConn = cc
+
+	// Initiate the trace service by sending over node identifier info.
+	traceSvcClient := agenttracepb.NewTraceServiceClient(cc)
+	traceExporter, err := traceSvcClient.Export(context.Background())
+	if err != nil {
+		return fmt.Errorf("Exporter.Start:: TraceServiceClient: %v", err)
+	}
+
+	firstTraceMessage := &agenttracepb.ExportTraceServiceRequest{Node: ae.nodeInfo}
+	err = nTriesWithExponentialBackoff(maxInitialTracesRetries, 200*time.Microsecond, func() error {
+		return traceExporter.Send(firstTraceMessage)
+	})
+	if err != nil {
+		return fmt.Errorf("Exporter.Start:: Failed to initiate the Config service: %v", err)
+	}
+	ae.traceExporter = traceExporter
+
+	// Initiate the config service by sending over node identifier info.
+	configStream, err := traceSvcClient.Config(context.Background())
+	if err != nil {
+		return fmt.Errorf("Exporter.Start:: ConfigStream: %v", err)
+	}
+	firstCfgMessage := &agenttracepb.CurrentLibraryConfig{Node: ae.nodeInfo}
+	err = nTriesWithExponentialBackoff(maxInitialConfigRetries, 200*time.Microsecond, func() error {
+		return configStream.Send(firstCfgMessage)
+	})
+	if err != nil {
+		return fmt.Errorf("Exporter.Start:: Failed to initiate the Config service: %v", err)
+	}
+
+	// In the background, handle trace configurations that are beamed down
+	// by the agent, but also reply to it with the applied configuration.
+	go ae.handleConfigStreaming(configStream)
+
+	return nil
+}
+
+func (ae *Exporter) handleConfigStreaming(configStream agenttracepb.TraceService_ConfigClient) error {
+	for {
```

Comment:
>In the future we can consider applying the config update asynchronously, so that if the agent sends a bunch of configs over a short period of time

I don't follow what you mean here. Do you mean that we should wait for some buffer period before applying the config? If so, I'd say I think we want to apply the config ASAP because any buffering/lag between applying a config could mean time lost and important events missed for a user e.g. during a traffic/latency spike.

Reply:
OK, make sense to me. I was thinking about when agent sent a wrong config and then sent another correct one for fixing up, in that case it may make sense to wait some buffer time, but those cases should be rare.